### ISSUE TITLE
Bug in _create_aggregators() with current bleeding edge theano

### ISSUE THREAD

**Issue by o1lo01ol1o** [mila-udem/blocks#820]
Updated theano on 8/28/2015 and now it complains about branch types in an ifelse:

```
  File "c:\Python27\lib\site-packages\blocks\extensions\monitoring.py", line 114
, in __init__
    self._buffer = AggregationBuffer(variables, use_take_last=True)
  File "c:\Python27\lib\site-packages\blocks\monitoring\evaluators.py", line 122
, in __init__
    self._create_aggregators()
  File "c:\Python27\lib\site-packages\blocks\monitoring\evaluators.py", line 147
, in _create_aggregators
    aggregator = v.tag.aggregation_scheme.get_aggregator()
  File "c:\Python27\lib\site-packages\blocks\monitoring\aggregation.py", line 11
8, in get_aggregator
    numerator_zeros)
  File "c:\Python27\lib\site-packages\theano\ifelse.py", line 367, in ifelse
    (then_branch_elem.type, else_branch_elem.type))
TypeError: The two branches should have identical types, but they are TensorType
(float32, vector) and TensorType(float32, (True,)) respectively. This error coul
d be raised if for example you provided a one element list on the `then` branch
but a tensor on the `else` branch.
```

**Comment by rizar** [comment]
What code do you run? Have you tried running standard examples?

**Comment by o1lo01ol1o** [comment]
I'm running some tandem LSTMs with feedward outputs; if you can point me to some standard examples I'll run them now and post the results.

**Comment by rizar** [comment]
Please run examples from https://github.com/mila-udem/blocks-examples.

**Comment by o1lo01ol1o** [comment]
The markov_chain test completes without a problem.  Unfortunately the prepare data script in the machine_translation directory breaks:

```
$ python machine_translation/prepare_data.py
INFO:prepare_data:Downloading [http://www.statmt.org/wmt15/training-parallel-nc-
v10.tgz]
INFO:prepare_data:...saving to: ./data\tmp\train_data.tgz Bytes: 125136590
INFO:prepare_data:Extracting file [./data\tmp\train_data.tgz] into [./data\tmp]
INFO:prepare_data:...extracting [news-commentary-v10.cs-en.cs] into [./data\tmp\
news-commentary-v10.cs-en.cs]
INFO:prepare_data:...extracting [news-commentary-v10.cs-en.en] into [./data\tmp\
news-commentary-v10.cs-en.en]
INFO:prepare_data:Downloading [http://www.statmt.org/wmt15/dev-v2.tgz]
INFO:prepare_data:...saving to: ./data\tmp\valid_data.tgz Bytes: 21393583
INFO:prepare_data:Extracting file [./data\tmp\valid_data.tgz] into [./data\tmp]
INFO:prepare_data:...extracting [dev/newstest2013.en] into [./data\tmp\dev/newst
est2013.en]
INFO:prepare_data:...extracting [dev/newstest2013.cs] into [./data\tmp\dev/newst
est2013.cs]
INFO:prepare_data:Downloading [https://raw.githubusercontent.com/moses-smt/moses
decoder/master/scripts/generic/multi-bleu.perl]
INFO:prepare_data:...saving to: ./data\multi-bleu.perl Bytes: 4576
INFO:prepare_data:Downloading [https://raw.githubusercontent.com/lisa-groundhog/
GroundHog/master/experiments/nmt/preprocess/preprocess.py]
INFO:prepare_data:...saving to: ./data\preprocess.py Bytes: 11479
INFO:prepare_data:Downloading [https://raw.githubusercontent.com/moses-smt/moses
decoder/master/scripts/tokenizer/tokenizer.perl]
INFO:prepare_data:...saving to: ./data\tokenizer.perl Bytes: 16587
INFO:prepare_data:Downloading [https://raw.githubusercontent.com/moses-smt/moses
decoder/master/scripts/share/nonbreaking_prefixes/nonbreaking_prefix.cs]
INFO:prepare_data:...saving to: ./share/nonbreaking_prefixes\nonbreaking_prefix.
cs Bytes: 1892
INFO:prepare_data:Downloading [https://raw.githubusercontent.com/moses-smt/moses
decoder/master/scripts/share/nonbreaking_prefixes/nonbreaking_prefix.en]
INFO:prepare_data:...saving to: ./share/nonbreaking_prefixes\nonbreaking_prefix.
en Bytes: 1327
INFO:prepare_data:Tokenizing file [./data\tmp\news-commentary-v10.cs-en.cs]
INFO:prepare_data:...writing tokenized file [./data\news-commentary-v10.cs-en.cs
.tok]

Traceback (most recent call last):
  File "machine_translation/prepare_data.py", line 258, in <module>
    main()
  File "machine_translation/prepare_data.py", line 242, in main
    tokenize_text_files(tr_files + val_files, tokenizer_file)
  File "machine_translation/prepare_data.py", line 118, in tokenize_text_files
    subprocess.Popen(var, stdin=inp, stdout=out, shell=False)
  File "c:\Python27\lib\subprocess.py", line 710, in __init__
    errread, errwrite)
  File "c:\Python27\lib\subprocess.py", line 958, in _execute_child
    startupinfo)
WindowsError: [Error 2] The system cannot find the file specified
```
This is probably a path issue on MINGW but I don't have the time right now to sort it out.  I'll look at running the reverse_words dataset next week and post back with an update.

**Comment by o1lo01ol1o** [comment]
I think I've found the issue.  The outputs from a `Missclassification().apply()` call were being flattened and not averaged before being passed to dataset monitor.  (Whoops, my bad.)  This was causing the training thread to become unresponsive or crash (presumably because `Aggregator` was expecting something saner).  As a debugging measure, I updated Theano, which then helpfully raised the `ifelse` violation before it exploded.  

So this issue can be closed, although maybe an assert check could be added to the class that calls the `Aggregator` just in case someone else replicates my mistake.
